src/team_lib/mhba/xccl_mhba_mkeys.c

```diff
@@ -179,55 +161,123 @@ static xccl_status_t poll_umr_cq(xccl_mhba_context_t *ctx) {
 // Execute the UMR WQE for populating the UMR's MasterMKey
 static xccl_status_t
 populate_mkey(xccl_mhba_context_t *ctx, xccl_mhba_node_t *node, int mem_access_flags, struct mlx5dv_mkey *mkey,
-              void *mkey_entries) {
+              void *mkey_entries, int block_size, int team_size) {
     xccl_status_t status;
-    xccl_mhba_debug("Execute the UMR WQE for populating the MasterMKeys");
     ibv_wr_start(ctx->umr_qpx);
     ctx->umr_qpx->wr_id = 1; // First (and only) WR todo: check meaning
-    int repeat_count = node->sbgp->group_size / node->block_size; // todo check calc
-    mlx5dv_wr_mr_interleaved(ctx->umr_mlx5dv_qp_ex, mkey, mem_access_flags,
-                             repeat_count, node->sbgp->group_size,
-                             (struct mlx5dv_mr_interleaved *) mkey_entries); // todo check for error
+    if (block_size) {
+        int repeat_count = team_size / block_size;
+        mlx5dv_wr_mr_interleaved(ctx->umr_mlx5dv_qp_ex, mkey, mem_access_flags,
+                                 repeat_count, node->sbgp->group_size,
+                                 (struct mlx5dv_mr_interleaved *) mkey_entries); // todo check for error
+        xccl_mhba_debug("Execute the UMR WQE for populating the send/recv MasterMKey lkey 0x%x",mkey->lkey);
+    } else{
+        mlx5dv_wr_mr_list(ctx->umr_mlx5dv_qp_ex, mkey, mem_access_flags, MAX_CONCURRENT_OUTSTANDING_ALL2ALL, (struct ibv_sge*)mkey_entries);
+        xccl_mhba_debug("Execute the UMR WQE for populating the team MasterMKeys lkey 0x%x",mkey->lkey);
+    }
+    ibv_wr_complete(ctx->umr_qpx);
     status = poll_umr_cq(ctx);
     if (status != XCCL_OK) {
         return status;
     }
     return XCCL_OK;
 }
 
-xccl_status_t xccl_mhba_populate_send_recv_mkeys(xccl_mhba_context_t *ctx, xccl_mhba_node_t *node){
+static xccl_status_t create_and_populate_team_mkey(xccl_mhba_context_t *ctx,xccl_mhba_node_t *node,struct mlx5dv_mkey** mkey,
+        struct mlx5dv_mkey** mkeys_array,int mem_access_flags,int team_size){
+    xccl_status_t status;
+    int i;
+    status = create_master_key(ctx, mkey, MAX_CONCURRENT_OUTSTANDING_ALL2ALL);
+    if (status != XCCL_OK) {
+        return status;
+    }
+    struct ibv_sge* team_mkey_klm_entries = (struct ibv_sge*)calloc(MAX_CONCURRENT_OUTSTANDING_ALL2ALL, sizeof(struct ibv_sge)); //todo destroy
+    for (i = 0; i < MAX_CONCURRENT_OUTSTANDING_ALL2ALL; i++) {
+        team_mkey_klm_entries[i].addr = 0;
+        team_mkey_klm_entries[i].length = node->sbgp->group_size*MAX_MSG_SIZE*team_size;
+        if(mem_access_flags) {
+            team_mkey_klm_entries[i].lkey = ((struct mlx5dv_mkey *) (*(mkeys_array + i)))->rkey;
```

Comment:
again, using "mem_access_flags" to distinguish between lkey and rkey is too opaque. need to have clear flags/behavior.